lessons/4.mdx

```diff
@@ -0,0 +1,818 @@
+
+If any of these functions are broken, you may not be able to mint NFTs, render
+the correct art or withdraw funds at the end.
+
+We want to test that those functions work in a variety of cases.
+
+For example, we can test that a mint happens when enough Eth is spent and that a
+mint fails if there is not enough Eth spent.
+
+### Start testing constructor(), mint() and withdraw()
+
+Let's start with the `constructor`. This one is pretty straight forward since
+the heavy lifting is done by OpenZeppelin's base classes, but let's make sure we
+can set the NFT name and symbol.
+
+Let's take a leap forward with our first two tests, as well as some setup code
+that we'll reuse for all the other tests too.
+
+```javascript
+const { expect } = require('chai')
+
+const CONTRACT_NAME = 'TierNFT'
+const COLLECTION_NAME = 'TierNFT'
+const COLLECTION_SYMBOL = 'Tier'
+
+describe('TierNFT', function () {
+  let contract
+  let owner
+  let otherUser
+
+  beforeEach(async function () {
+    const Contract = await hre.ethers.getContractFactory(CONTRACT_NAME)
+
+    const [_owner, _otherUser] = await hre.ethers.getSigners()
+    owner = _owner
+    otherUser = _otherUser
+
+    contract = await Contract.deploy(COLLECTION_NAME, COLLECTION_SYMBOL)
+    await contract.deployed()
+  })
+
+  describe('constructor', async () => {
+    it('set proper collection name', async function () {
+      const name = await contract.name()
+      expect(name).to.equal('TierNFT')
+    })
+
+    it('set proper collection symbol', async function () {
+      const symbol = await contract.symbol()
+      expect(symbol).to.equal('Tier')
+    })
+  })
+})
+```
+
+At the top we see constants with the name and symbol of the NFT collection:
+
+```javascript
+const CONTRACT_NAME = 'TierNFT'
+const COLLECTION_NAME = 'TierNFT'
+const COLLECTION_SYMBOL = 'Tier'
+```
+
+Next is a `describe('TierNFT', function () { ...` line. What's cool about
+`describe()` is that it logically groups tests together so you'll see nicely
+organized output.
+
+When we run the tests we see:
+
+```bash
+TierNFT
+    constructor
+      ✔ set proper collection name
+      ✔ set proper collection symbol
+```
+
+Notice "TierNFT" comes from the first `describe()` block and `constructor` is
+from the inner `describe()` block. The tests are then indented under there.
+Makes for a nicely organized suite of tests.
+
+Next notice the `beforeEach()` code block. When tests run it's important that
+the contract is deployed from scratch for each test and this is where that
+happens. "Before each" test, this code will run and re-inialize everything,
+which prevents one test's results from interfering with another test.
+
+What's happening in `beforeEach()`?
+
+The hardhat framework give us the tools to deploy our contract to a
+locally-running blockchain. So when this code finishes, we have a fully deployed
+contract, and a couple of wallets to use for testing -- `owner` is who has all
```

Comment:
a couple of wallets to use for testing including `owner` is who has all the permissions, since it deployed the contract, and `otherUser` is a wallet with fewer permissions. These will be handy when testing `withdraw` later on.